images/providers.go

```diff
@@ -0,0 +1,100 @@
+package images
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/models"
+	"github.com/prometheus/alertmanager/types"
+)
+
+// ErrImageUploadNotSupported is returned when image uploading is not supported.
+var ErrImageUploadNotSupported = errors.New("image upload is not supported")
+
+type UnavailableProvider struct{}
+
+var _ Provider = (*UnavailableProvider)(nil)
+
+func (u *UnavailableProvider) GetImage(_ context.Context, _ types.Alert) (*Image, error) {
+	return nil, ErrImagesUnavailable
+}
+
+// URLProvider is a provider that stores a direct reference to an image's public URL in an alert's annotations.
+// The URL is not validated against a database record, so retrieving raw image data is blocked in an attempt
+// to prevent malicious access to untrusted URLs.
+type URLProvider struct{}
+
+var _ Provider = (*URLProvider)(nil)
+
+// GetImage returns the image associated with a given alert.
+// The URL should be treated as untrusted and notifiers should pass the URL directly without attempting to download
+// the image data.
+func (u *URLProvider) GetImage(_ context.Context, alert types.Alert) (*Image, error) {
+	url := GetImageURL(alert)
+	if url == "" {
+		return nil, nil
+	}
+
+	return &Image{
+		URL: url,
+		RawData: func(_ context.Context) (ImageContent, error) {
+			// Raw images are not available for alerts provided directly by annotations as the image data is non-local.
+			// While it might be possible to download the image data, it's generally not safe to do so as the URL is
+			// not guaranteed to be trusted.
+			return ImageContent{}, ErrImageUploadNotSupported
+		},
+	}, nil
+}
+
+type TokenStore interface {
+	GetImage(ctx context.Context, token string) (*Image, error)
+}
+
+// TokenProvider is an implementation of the Image Provider interface that retrieves images from a store
+// using tokens.
+// Image data should be considered trusted by the fact that the stored image url and content and not user-modifiable.
+type TokenProvider struct {
+	store  TokenStore
+	logger logging.Logger
+}
+
+var _ Provider = (*TokenProvider)(nil)
+
+func NewTokenProvider(store TokenStore, logger logging.Logger) Provider {
+	return &TokenProvider{
+		store:  store,
+		logger: logger,
+	}
+}
+
+func (i TokenProvider) GetImage(ctx context.Context, alert types.Alert) (*Image, error) {
+	token := GetImageToken(alert)
+	if token == "" {
+		return nil, nil
+	}
+
+	// Assume the uri is a token because we used to store tokens as plain strings.
+	logger := i.logger.New("token", token)
+	logger.Debug("Received an image token in annotations")
```

Comment:
[`logger.New()`](https://github.com/grafana/grafana/blob/7c8fddc7291f181b83fee6e6a7be4f1190a38229/pkg/infra/log/log.go#L175-L189) creates a new logger struct and returns a pointer. This reference to the logger will escape to the heap and need to be GC'd. Some other allocations are done along the way, as the function also calls some methods from 3rd party libraries.

I would add the context to the original logger and use it.
```suggestion
	i.logger.Debug("Received an image token in annotations", "token", token)
```

Same for the `Info` line below.